### PR TITLE
zig cc improvements

### PR DESCRIPTION
**Pull request by vrischmann**
This PR improves the linker args parsing for `zig cc`.

### CODE REVIEW

**Comment by mikdusan** [comment]
Rather than handle each `-Wl,-OPTION=ARG` explicitly, how about generically handle any `.comma_joined` by detecting the `-OPTION=ARG` pattern and add 2 arguments to linker_args for further processing?

```diff
diff --git a/src/main.zig b/src/main.zig
index 512828bd7..8da878888 100644
--- a/src/main.zig
+++ b/src/main.zig
@@ -1179,6 +1179,15 @@ fn buildOutputType(
                     .wl => {
                         var split_it = mem.split(it.only_arg, ",");
                         while (split_it.next()) |linker_arg| {
+                            // Handle nested-joined args like `-Wl,-rpath=foo` .
+                            // Must be prefixed with 1 or 2 dashes.
+                            if (linker_arg.len >= 3 and linker_arg[0] == '-' and linker_arg[2] != '-') {
+                                if (mem.indexOfScalar(u8, linker_arg, '=')) |equals_pos| {
+                                    try linker_args.append(linker_arg[0..equals_pos]);
+                                    try linker_args.append(linker_arg[equals_pos + 1..]);
+                                    continue;
+                                }
+                            }
                             try linker_args.append(linker_arg);
                         }
                     },
```

**Comment by vrischmann** [comment]
I wasn't really sure if it would be accepted to alter how `linker_args` is generated here, but ok I can do that.

**Review by mikdusan — APPROVED** [review]
Real nice that you clipped-out previous specialized handling of linker `-option=`.

File: src/main.zig [hunk 1]
```diff
@@ -1179,6 +1179,16 @@ fn buildOutputType(
...
                         while (split_it.next()) |linker_arg| {
+                            // Handle nested-joined args like `-Wl,-rpath=foo`.
+                            // Must be prefixed with 1 or 2 dashes.
+                            if (linker_arg.len >= 3 and linker_arg[0] == '-' and linker_arg[2] != '-') {
+                                if (mem.indexOfScalar(u8, linker_arg, '=')) |equals_pos| {
+                                    try linker_args.append(linker_arg[0..equals_pos]);
+                                    try linker_args.append(linker_arg[equals_pos + 1 ..]);
+                                    continue;
+                                }
+                            }
```
Comment (andrewrk) [inline src/main.zig:1190]: Are you sure this is correct? Seems weird that it's asking to pass `-rpath=foo` to the linker, but instead we pass `-rpath foo`

File: src/main.zig [hunk 2]
```diff
@@ -1420,30 +1430,40 @@ fn buildOutputType(
                     // We don't need to care about these because these args are
                     // for resolving circular dependencies but our linker takes
                     // care of this without explicit args.
-                } else if (mem.startsWith(u8, arg, "--major-os-version") or
-                    mem.startsWith(u8, arg, "--minor-os-version"))
+                } else if (mem.eql(u8, arg, "--major-os-version") or
+                    mem.eql(u8, arg, "--minor-os-version"))
```
Comment (andrewrk) [inline src/main.zig:1434]: This breaks `--major-os-version=foo`

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: src/main.zig (hunk 1 above)
Comment (mikdusan) [reply src/main.zig:1190]: This is correct. Both macos and linux system linkers accepts BOTH 2-arg and 1-arg-with-equals syntax. Ultimately, we choose the easiest way for our linker-args pass to just see it as 2-arg (which it already supported before this PR).

**Comment by andrewrk** [comment]
Note that `-Wl` is not the only way linker_args gets appended to. There is also --for-linker and -z

**Comment by andrewrk** [comment]
For linker args I'd encourage pushing on #8755 by moving towards a data set of known flags and mappings to zig concepts, same as how we do for clang command line arguments with the tools/update_clang_options.zig.

Minor improvements along the way to our existing linker flag parsing like this are fine as long as they don't have any regressions, but I spotted a problem here.

File: src/main.zig (hunk 2 above)
Comment (mikdusan) [reply src/main.zig:1434]: I don't understand. Previous to PR these options were no-op but `startsWith` creates some problems:
- accepts `--major-os-version`, `--major-os-version=`, and `--major-os-versionANYTHING`
- if `--major-os-version` is used it is likely the user read the man page and used 2-arg syntax, and in that case following arg is not consumed

afaict this args don't do anything anyways and IMO we should accept them in both 2-arg and 1-arg-with-equals syntaxes to properly eat the command line args.

**Comment by mikdusan** [comment]
On IRC we resolved the outstanding concerns.

**Pull request merged**